In [21]:
import pandas as pd

## 数据预处理
首先我们对数据进行预处理操作。

### 缺省值处理
我们首先处理所有的`unknown`的值

In [22]:
data = pd.read_csv('./dataset/bank-additional-full.csv', sep=";")

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

我们可以看出数据一共有41188条，虽然没有缺省值，但是其中还是有不少的`unknown`的数据，我们首先查看这些数据的数量。

In [24]:
def calculate_unknown(data, attributes, target="unknown"):
    for att in attributes:
        print("The missing number of "+att+" is "+str(len(data[data[att]==target])))

In [25]:
calculate_unknown(data, list(data.columns))

The missing number of age is 0
The missing number of job is 330
The missing number of marital is 80
The missing number of education is 1731
The missing number of default is 8597
The missing number of housing is 990
The missing number of loan is 990
The missing number of contact is 0
The missing number of month is 0
The missing number of day_of_week is 0
The missing number of duration is 0
The missing number of campaign is 0
The missing number of pdays is 0
The missing number of previous is 0
The missing number of poutcome is 0
The missing number of emp.var.rate is 0
The missing number of cons.price.idx is 0
The missing number of cons.conf.idx is 0
The missing number of euribor3m is 0
The missing number of nr.employed is 0
The missing number of y is 0


从上述的结果可以看出，对于总数41188而言，`unknown`的占比不是非常大，所以在本次作业中我们可以直接将缺省值删去。

In [26]:
def delete_unknown(data):
    for att in list(data.columns):
        data = data[data[att]!="unknown"]
    return data

In [27]:
data = delete_unknown(data)

In [28]:
len(data)

30488

目前还剩下30488的数据，数据量还是比较可观的。

In [29]:
data.head(20)

age          job   marital            education default housing loan  \
0    56    housemaid   married             basic.4y      no      no   no   
2    37     services   married          high.school      no     yes   no   
3    40       admin.   married             basic.6y      no      no   no   
4    56     services   married          high.school      no      no  yes   
6    59       admin.   married  professional.course      no      no   no   
8    24   technician    single  professional.course      no     yes   no   
9    25     services    single          high.school      no     yes   no   
11   25     services    single          high.school      no     yes   no   
12   29  blue-collar    single          high.school      no      no  yes   
13   57    housemaid  divorced             basic.4y      no     yes   no   
14   35  blue-collar   married             basic.6y      no     yes   no   
16   35  blue-collar   married             basic.6y      no     yes   no   
18   50  blue-collar   married             basic.9y      no     yes  yes   
20   30   unemployed   married          high.school      no      no   no   
22   55      retired    single          high.school      no     yes   no   
23   41   technician    single          high.school      no     yes   no   
24   37       admin.   married          high.school      no     yes   no   
25   35   technician   married    university.degree      no      no  yes   
34   54  blue-collar  divorced             basic.4y      no      no   no   
36   34     services   married          high.school      no      no   no   

      contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0   telephone   may         mon  ...         1    999         0  nonexistent   
2   telephone   may         mon  ...         1    999         0  nonexistent   
3   telephone   may         mon  ...         1    999         0  nonexistent   
4   telephone   may         mon  ...         1    999         0  nonexistent   
6   telephone   may         mon  ...         1    999         0  nonexistent   
8   telephone   may         mon  ...         1    999         0  nonexistent   
9   telephone   may         mon  ...         1    999         0  nonexistent   
11  telephone   may         mon  ...         1    999         0  nonexistent   
12  telephone   may         mon  ...         1    999         0  nonexistent   
13  telephone   may         mon  ...         1    999         0  nonexistent   
14  telephone   may         mon  ...         1    999         0  nonexistent   
16  telephone   may         mon  ...         1    999         0  nonexistent   
18  telephone   may         mon  ...         1    999         0  nonexistent   
20  telephone   may         mon  ...         1    999         0  nonexistent   
22  telephone   may         mon  ...         1    999         0  nonexistent   
23  telephone   may         mon  ...         1    999         0  nonexistent   
24  telephone   may         mon  ...         1    999         0  nonexistent   
25  telephone   may         mon  ...         1    999         0  nonexistent   
34  telephone   may         mon  ...         1    999         0  nonexistent   
36  telephone   may         mon  ...         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0           1.1          93.994          -36.4      4.857       5191.0  no  
2           1.1          93.994          -36.4      4.857       5191.0  no  
3           1.1          93.994          -36.4      4.857       5191.0  no  
4           1.1          93.994          -36.4      4.857       5191.0  no  
6           1.1          93.994          -36.4      4.857       5191.0  no  
8           1.1          93.994          -36.4      4.857       5191.0  no  
9           1.1          93.994          -36.4      4.857       5191.0  no  
11          1.1          93.994          -36.4      4.857       5191.0  no  
12          1.1          93.994  

我们需将data中的values都转换为str类型，方便后续操作。

In [30]:
data = data.astype(str)

通过观察value，我们不难看出`default`, `housing`, `loan`, and `y`这几个column的值是相同的，所以我们要将他们区别开来。
为了之后研究关联规则更加方便，我们将所有的值都加上其column的前缀

In [31]:
def diff(data, columns):
    for column in columns:
        temp = list(data[column])
        for i in range(len(temp)):
            temp[i] = column+"_"+temp[i]
        data[column] = temp
    return data

In [32]:
data = diff(data, list(data.columns))

In [33]:
data

age              job          marital  \
0      age_56    job_housemaid  marital_married   
2      age_37     job_services  marital_married   
3      age_40       job_admin.  marital_married   
4      age_56     job_services  marital_married   
6      age_59       job_admin.  marital_married   
...       ...              ...              ...   
41183  age_73      job_retired  marital_married   
41184  age_46  job_blue-collar  marital_married   
41185  age_56      job_retired  marital_married   
41186  age_44   job_technician  marital_married   
41187  age_74      job_retired  marital_married   

                           education     default      housing      loan  \
0                 education_basic.4y  default_no   housing_no   loan_no   
2              education_high.school  default_no  housing_yes   loan_no   
3                 education_basic.6y  default_no   housing_no   loan_no   
4              education_high.school  default_no   housing_no  loan_yes   
6      education_professional.course  default_no   housing_no   loan_no   
...                              ...         ...          ...       ...   
41183  education_professional.course  default_no  housing_yes   loan_no   
41184  education_professional.course  default_no   housing_no   loan_no   
41185    education_university.degree  default_no  housing_yes   loan_no   
41186  education_professional.course  default_no   housing_no   loan_no   
41187  education_professional.course  default_no  housing_yes   loan_no   

                 contact      month      day_of_week  ...    campaign  \
0      contact_telephone  month_may  day_of_week_mon  ...  campaign_1   
2      contact_telephone  month_may  day_of_week_mon  ...  campaign_1   
3      contact_telephone  month_may  day_of_week_mon  ...  campaign_1   
4      contact_telephone  month_may  day_of_week_mon  ...  campaign_1   
6      contact_telephone  month_may  day_of_week_mon  ...  campaign_1   
...                  ...        ...              ...  ...         ...   
41183   contact_cellular  month_nov  day_of_week_fri  ...  campaign_1   
41184   contact_cellular  month_nov  day_of_week_fri  ...  campaign_1   
41185   contact_cellular  month_nov  day_of_week_fri  ...  campaign_2   
41186   contact_cellular  month_nov  day_of_week_fri  ...  campaign_1   
41187   contact_cellular  month_nov  day_of_week_fri  ...  campaign_3   

           pdays    previous              poutcome       emp.var.rate  \
0      pdays_999  previous_0  poutcome_nonexistent   emp.var.rate_1.1   
2      pdays_999  previous_0  poutcome_nonexistent   emp.var.rate_1.1   
3      pdays_999  previous_0  poutcome_nonexistent   emp.var.rate_1.1   
4      pdays_999  previous_0  poutcome_nonexistent   emp.var.rate_1.1   
6      pdays_999  previous_0  poutcome_nonexistent   emp.var.rate_1.1   
...          ...         ...                   ...                ...   
41183  pdays_999  previous_0  poutcome_nonexistent  emp.var.rate_-1.1   
41184  pdays_999  previous_0  poutcome_nonexistent  emp.var.rate_-1.1   
41185  pdays_999  previous_0  poutcome_nonexistent  emp.var.rate_-1.1   
41186  pdays_999  previous_0  poutcome_nonexistent  emp.var.rate_-1.1   
41187  pdays_999  previous_1      poutcome_failure  emp.var.rate_-1.1   

              cons.price.idx        cons.conf.idx        euribor3m  \
0      cons.price.idx_93.994  cons.conf.idx_-36.4  euribor3m_4.857   
2      cons.price.idx_93.994  cons.conf.idx_-36.4  euribor3m_4.857   
3      cons.price.idx_93.994  cons.conf.idx_-36.4  euribor3m_4.857   
4      cons.price.idx_93.994  cons.conf.idx_-36.4  euribor3m_4.857   
6      cons.price.idx_93.994  cons.conf.idx_-36.4  euribor3m_4.857   
...                      ...                  ...              ...   
41183  cons.price.idx_94.767  cons.conf.idx_-50.8  euribor3m_1.028   
41184  cons.price.idx_94.767  cons.conf.idx_-50.8  euribor3m_1.028   
41185  cons.price.idx_94.767  cons.conf.idx_-50.8  euribor3m_1.028   
41186  cons.price.idx_94.767  cons.conf.id

将data转换为list形式。

In [34]:
data_set = data.values.tolist()

## 使用apriori找出频繁模式
本实验使用的apriori的具体代码都在`apriori.py`和`utilis.py`中可以找到。

### Apriori算法介绍
在介绍Apriori算法之前，我们需要先明晰两个概念。
#### Support
$$\begin{equation}
\operatorname{supp}(X)=\frac{\text { Number of transaction in which } X \text { appears }}{\text { Total number of transactions }}
\end{equation}$$
#### Confidence
$$\begin{equation}
\operatorname{con} f(X \longrightarrow Y)=\frac{\operatorname{supp}(X \cup Y)}{\operatorname{supp}(X)}
\end{equation}$$
#### Apriori目的
对于Apriori算法，我们使用支持度来作为我们判断频繁项集的标准。Apriori算法的目标是找到最大的K项频繁集。
#### Apriori执行步骤
Apriori算法采用了迭代的方法，先搜索出候选1项集及对应的支持度，剪枝去掉低于支持度阙值的1项集，得到频繁1项集。然后对剩下的频繁1项集进行连接，得到频繁2项集，筛选去掉低于支持度阙值的候选2项集，得到真正的频繁2项集。以此类推，迭代下去，直到无法找到频繁k+1项集为止，对应的频繁k项集的集合即为算法的输出结果。
#### 关联规则的生成
1. 对于每个频繁项集L，生成其所有的非空子集
2. 对于L的每个非空子集x, 计算其置信度Confidence（x）≥ minConfidence，那么“ x -> (L-x)”成立。

In [44]:
from apriori import apriori, association
# 获取candidate和support
# 设置mini_support为0.5
l, support = apriori(data_set, 0.5)

In [54]:
# 生成关联规则
# 设置mini_conf为0.9
rules = association(l, support, 0.9)

In [55]:
# 按照conf从大到小排序
rules.sort(key=lambda x: x[-1])

In [56]:
for rule in rules:
    print(str(rule[0])+"-->"+str(rule[1])+" and the conf: "+str(rule[-1]))

frozenset({'poutcome_nonexistent', 'loan_no', 'default_no'})-->frozenset({'y_no'}) and the conf: 0.9000045890505254
frozenset({'previous_0', 'loan_no', 'default_no'})-->frozenset({'y_no'}) and the conf: 0.9000045890505254
frozenset({'poutcome_nonexistent', 'default_no', 'loan_no', 'pdays_999'})-->frozenset({'y_no'}) and the conf: 0.9000045890505254
frozenset({'previous_0', 'default_no', 'loan_no', 'pdays_999'})-->frozenset({'y_no'}) and the conf: 0.9000045890505254
frozenset({'poutcome_nonexistent', 'previous_0', 'default_no', 'loan_no'})-->frozenset({'y_no'}) and the conf: 0.9000045890505254
frozenset({'default_no', 'loan_no', 'pdays_999', 'poutcome_nonexistent', 'previous_0'})-->frozenset({'y_no'}) and the conf: 0.9000045890505254
frozenset({'previous_0', 'loan_no'})-->frozenset({'y_no'}) and the conf: 0.9000137658881293
frozenset({'poutcome_nonexistent', 'loan_no'})-->frozenset({'y_no'}) and the conf: 0.9000137658881293
frozenset({'previous_0', 'loan_no', 'pdays_999'})-->frozenset({

## 关联规则分析
### column的具体含义
我们在该数据集的描述中可以找到列的含义，大致如下：
<img src="./assets/attributes.png">


通过对上述的关联规则的分析，我们发现生成的关联规则都是最终导向"y_no"。其实这是因为我们设置了`mini_support`为0.5之后，"y_yes"的项再生成1-项集的时候就被去除了。

由于篇幅的原因，我们只对confidence最大的前五条关联规则进行说明解释：
> frozenset({'loan_no', 'poutcome_nonexistent', 'default_no'})-->frozenset({'y_no'})

> frozenset({'previous_0', 'loan_no', 'default_no'})-->frozenset({'y_no'})

> frozenset({'loan_no', 'default_no', 'poutcome_nonexistent', 'pdays_999'})-->frozenset({'y_no'})

> frozenset({'loan_no', 'default_no', 'previous_0', 'poutcome_nonexistent'})-->frozenset({'y_no'})

> frozenset({'loan_no', 'default_no', 'previous_0', 'pdays_999'})-->frozenset({'y_no'})





通过上述对column的解读，我们会发现其中loan_no和defautl_no对与该用户没有定期存款的影响非常大。我们会发现其中`pdays_999`的出镜率也非常高，但是通过对`pdays`的数值进行分析会发现，`pdays`的值的分布非常不均衡：

In [57]:
data.groupby('pdays')['pdays'].value_counts()

pdays      pdays    
pdays_0    pdays_0         14
pdays_1    pdays_1         21
pdays_10   pdays_10        40
pdays_11   pdays_11        25
pdays_12   pdays_12        50
pdays_13   pdays_13        33
pdays_14   pdays_14        17
pdays_15   pdays_15        22
pdays_16   pdays_16         8
pdays_17   pdays_17         6
pdays_18   pdays_18         5
pdays_19   pdays_19         3
pdays_2    pdays_2         53
pdays_21   pdays_21         2
pdays_22   pdays_22         3
pdays_25   pdays_25         1
pdays_26   pdays_26         1
pdays_27   pdays_27         1
pdays_3    pdays_3        381
pdays_4    pdays_4        102
pdays_5    pdays_5         43
pdays_6    pdays_6        363
pdays_7    pdays_7         50
pdays_8    pdays_8         13
pdays_9    pdays_9         53
pdays_999  pdays_999    29178
Name: pdays, dtype: int64

其中高达29178的数据都是`pdays_999`，这样极度不均衡的数据导致`pdays`的项出现在最后的结果中也不足为奇，所以我们在分析中可以直接将`pdays`这个项给略去。

在分析完关联规则之后，为了分析更加直观和方便，我们将生成的强关联规则进行整理之后分析单个项对于结果的影响：

In [59]:
count = {}
for rule in rules:
    for item in rule[0]:
        if item not in count:
            count[item] = 1
        else:
            count[item] += 1

In [64]:
sorted(count, key=lambda x: x[-1], reverse=True)

['poutcome_nonexistent', 'loan_no', 'default_no', 'pdays_999', 'previous_0']

通过分析我们可以发现`poutcome`和个人贷款以及个人信用都是和该用户是否进行定期存款比较相关的项。

最后我们再将计算生成的强关联规则展示：

In [66]:
for rule in rules:
    print(str(rule[0])+"--->"+str(rule[1]))

frozenset({'poutcome_nonexistent', 'loan_no', 'default_no'})--->frozenset({'y_no'})
frozenset({'previous_0', 'loan_no', 'default_no'})--->frozenset({'y_no'})
frozenset({'poutcome_nonexistent', 'default_no', 'loan_no', 'pdays_999'})--->frozenset({'y_no'})
frozenset({'previous_0', 'default_no', 'loan_no', 'pdays_999'})--->frozenset({'y_no'})
frozenset({'poutcome_nonexistent', 'previous_0', 'default_no', 'loan_no'})--->frozenset({'y_no'})
frozenset({'default_no', 'loan_no', 'pdays_999', 'poutcome_nonexistent', 'previous_0'})--->frozenset({'y_no'})
frozenset({'previous_0', 'loan_no'})--->frozenset({'y_no'})
frozenset({'poutcome_nonexistent', 'loan_no'})--->frozenset({'y_no'})
frozenset({'previous_0', 'loan_no', 'pdays_999'})--->frozenset({'y_no'})
frozenset({'previous_0', 'loan_no', 'poutcome_nonexistent'})--->frozenset({'y_no'})
frozenset({'poutcome_nonexistent', 'loan_no', 'pdays_999'})--->frozenset({'y_no'})
frozenset({'previous_0', 'poutcome_nonexistent', 'loan_no', 'pdays_999'})--->fr